### ISSUE TITLE
TF-TRT slower than optimized saved model

### ISSUE THREAD

**Issue by jtressle** [tensorflow/tensorflow#32103]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, I have a network that does 2D convultions + batch normalization on an image.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): TF 1.14
- Python version: 3.6.8
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): 7.4
- CUDA/cuDNN version: 10.0
- GPU model and memory: T4, 12GB

**Describe the current behavior**

I'm trying to optimize a custom model comprised of 2D convolutions and batch normalizations done on an image.  The entire network has fixed dimensions.  I'm using the nightly docker TF image to perform TF-TRT.

I've tried to create a TRT model using both of the following functions:

```
def create_trt_saved_model(saved_model_dir, output_saved_model_dir, precision, batch_size=1):
	''' convert saved model to TRT saved model'''

	converter = trt.TrtGraphConverter(
		input_saved_model_dir = str(saved_model_dir),
		max_batch_size = batch_size,
		precision_mode = precision )
	converter.convert()
	converter.save(output_saved_model_dir = str(output_saved_model_dir))
```

and

```
def create_trt_frozen_graph(graph_def, output_nodes, precision, 
	output_graph_path = None, workspace_size=2<<10, batch_size=1):
	''' convert frozen_graph to a TRT frozen graph'''
	
	converter = trt.TrtGraphConverter(
		input_graph_def = graph_def,
		nodes_blacklist = output_nodes,
		max_batch_size = batch_size,
		max_workspace_size_bytes = workspace_size<<20,
		precision_mode = precision)

	trt_graph_def = converter.convert()

	if not (output_graph_path is None):
		write_graph_to_file(trt_graph_def, output_graph_path)

	return trt_graph_def
```

In both cases, the TF-TRT model is about 35X slower (20ms vs 700ms inference).  The results are the same regardless if I use the graph_def from memory, or load the TF-TRT saved model.

Here is the respective TRT output:

```
2019-08-30 04:48:35.865582: I tensorflow/compiler/tf2tensorrt/segment/segment.cc:460] There are 4 ops of 3 different types in the graph that are not converted to TensorRT: Identity, NoOp, Placeholder, (For more information see https://docs.nvidia.com/deeplearning/frameworks/tf-trt-user-guide/index.html#supported-ops).
2019-08-30 04:48:35.953878: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:633] Number of TensorRT candidate segments: 1
2019-08-30 04:48:35.969432: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libnvinfer.so.5
2019-08-30 04:48:35.969838: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libnvinfer_plugin.so.5
2019-08-30 04:55:04.225714: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:734] TensorRT node TRTEngineOp_0 added for segment 0 consisting of 944 nodes succeeded.
2019-08-30 04:55:04.352053: W tensorflow/compiler/tf2tensorrt/convert/trt_optimization_pass.cc:183] TensorRTOptimizer is probably called on funcdef! This optimizer must *NOT* be called on function objects.
2019-08-30 04:55:04.402986: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:786] Optimization results for grappler item: tf_graph
2019-08-30 04:55:04.403043: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 818 nodes (-817), 914 edges (-949), time = 62.326ms.
2019-08-30 04:55:04.403049: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   layout: Graph size after: 950 nodes (132), 1046 edges (132), time = 50.486ms.
2019-08-30 04:55:04.403054: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 946 nodes (-4), 1042 edges (-4), time = 45.175ms.
2019-08-30 04:55:04.403059: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   TensorRTOptimizer: Graph size after: 3 nodes (-943), 2 edges (-1040), time = 388396.844ms.
2019-08-30 04:55:04.403063: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 3 nodes (0), 2 edges (0), time = 2.443ms.
2019-08-30 04:55:04.403067: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:786] Optimization results for grappler item: TRTEngineOp_0_native_segment
2019-08-30 04:55:04.403072: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 946 nodes (0), 1042 edges (0), time = 26.303ms.
2019-08-30 04:55:04.403076: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   layout: Graph size after: 946 nodes (0), 1042 edges (0), time = 30.005ms.
2019-08-30 04:55:04.403092: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 946 nodes (0), 1042 edges (0), time = 26.387ms.
2019-08-30 04:55:04.403097: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   TensorRTOptimizer: Graph size after: 946 nodes (0), 1042 edges (0), time = 3.387ms.
2019-08-30 04:55:04.403103: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 946 nodes (0), 1042 edges (0), time = 26.708ms.
2019-08-30 04:55:04.727419: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at 
```

Since the TRT model only has 3 nodes, one of which is the TRT Engine node, does it make sense to convert this via UFF?  Would that get me a speed improvement?

Or, is there a bug in the latest version of TF docker?

Thanks!

**Describe the expected behavior**

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by pooyadavoodi** [comment]
I wonder why you are seeing `Optimization results for grappler item: TRTEngineOp_0_native_segment`. Typically only `Optimization results for grappler item: tf_graph` is reported. Looks like you are running the conversion twice?

**Comment by pooyadavoodi** [comment]
Are you using INT8 by any chance? If yes, you might be in the calibration phase which is much slower than inference. You would need to end the calibration. 

Here is the docs: https://docs.nvidia.com/deeplearning/frameworks/tf-trt-user-guide/index.html#post-train

**Comment by pooyadavoodi** [comment]
Note that TF-TRT has made one TRT segment (TRTEngineOp_0) that has 944 nodes in it. This is a large subgraph which is the ideal case.

**Comment by eddienewton** [comment]
@pooyadavoodi thanks for the replies.

I'm getting similar results using either FP16 and FP32.  I first wanted to test these before trying calibration.

I'm going straight from the saved model (created with following code) to the convert_trt code.  I did notice using saved_model_cli actually converts my saved model to TRT faster.

```
builder = tf.saved_model.builder.SavedModelBuilder(str(graph_path))

#define signatures
prediction_signature = (
	tf.saved_model.signature_def_utils.build_signature_def(
		inputs=inputs,
		outputs=outputs,
		method_name=tf.saved_model.signature_constants.PREDICT_METHOD_NAME))

builder.add_meta_graph_and_variables(
	sess, [tf.saved_model.tag_constants.SERVING],
	signature_def_map = 
		{tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY : prediction_signature},
	clear_devices=False)

builder.save(as_text=False)

writer = tf.summary.FileWriter("output", sess.graph)
writer.close()
```

I'll take a look to see how I could be running the conversion twice.  From what I remember, I'm simply specifying the saved model folder, and then running `create_trt_saved_model`.

Also, is this the correct link to start profiling?  

[https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#profiling](https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#profiling)

**Comment by jtressle** [comment]
@pooyadavoodi thanks for the replies.

I'm getting similar results using either FP16 and FP32. I first wanted to test these before trying calibration.

I'm going straight from the saved model (created with following code) to the convert_trt code. I did notice using saved_model_cli actually converts my saved model to TRT faster.

builder = tf.saved_model.builder.SavedModelBuilder(str(graph_path))

#define signatures
prediction_signature = (
	tf.saved_model.signature_def_utils.build_signature_def(
		inputs=inputs,
		outputs=outputs,
		method_name=tf.saved_model.signature_constants.PREDICT_METHOD_NAME))

builder.add_meta_graph_and_variables(
	sess, [tf.saved_model.tag_constants.SERVING],
	signature_def_map = 
		{tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY : prediction_signature},
	clear_devices=False)

builder.save(as_text=False)

writer = tf.summary.FileWriter("output", sess.graph)
writer.close()
I'll take a look to see how I could be running the conversion twice. From what I remember, I'm simply specifying the saved model folder, and then running create_trt_saved_model.

Also, is this the correct link to start profiling?

https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#profiling

**Comment by pooyadavoodi** [comment]
@jtressle the profiling link is correct.

**Comment by aaroey** [comment]
@jtressle TF 1.14 has TF-TRT disabled accidentally. Could you try again with TF 1.15.0rc1? Also could you provide the model data so we can reproduce the slowness?

**Comment by leo-XUKANG** [comment]
maybe i can answer the question, when the start a sess and load the model which has been optimiszed by TF-TRT, it will load the TRT-engine op , and it will use a lot of time, maybe some seconds,but you will get a smaller latency in the rest test

**Comment by aaroey** [comment]
@jtressle I think what @leo-XUKANG could be possible. I'm closing this, please feel free to reopen and provide the model and script to reproduce the problem if it still exists.
